### ISSUE TITLE
BUG: np.allclose complex numbers with a part np.inf

### ISSUE THREAD

**Issue by zerothi** [numpy/numpy#20291]
### Describe the issue:

Comparing two complex values with one of them having an infinite part (either real or complex) results in the comparison just failing.

This does _not_ happen for proper real `np.inf` values.

The below code snippet should in my view return all True's.

### Reproduce the code example:

```python
import numpy as np
print(np.allclose(1., 1.))
print(np.allclose(1., 1. + 0j))
print(np.allclose(np.inf, np.inf))
print(np.allclose(np.inf + 0j, np.inf))
print(np.allclose(np.inf + 0j, np.inf + 0j))
print(np.allclose(1 + np.inf * 1j, 1 + np.inf * 1j))
```

### Error message:

```shell
True
True
True
/opt/gnu/11.2.0/python/packages/3.9.7/numpy/1.21.3/lib/python3.9/site-packages/numpy/core/numeric.py:2365: RuntimeWarning: invalid value encountered in multiply
  x = x * ones_like(cond)
False
/opt/gnu/11.2.0/python/packages/3.9.7/numpy/1.21.3/lib/python3.9/site-packages/numpy/core/numeric.py:2366: RuntimeWarning: invalid value encountered in multiply
  y = y * ones_like(cond)
False
False
```

### NumPy/Python version information:

1.21.3 3.9.7 (default, Nov  2 2021, 13:08:39) 
[GCC 11.2.0]

**Comment by UmarJ** [comment]
This is because the function contains multiplication of a complex number having `np.inf` as one of its parts with a real number.

https://github.com/numpy/numpy/blob/c280e21a74f2a6fd16d9c318696fcf167689baa3/numpy/core/numeric.py#L2368-L2369

See #20332.

**Comment by zerothi** [comment]
Agreed, that is the reason for the error. But this means that allclose should check for inf in complex cases before using mult...

**Comment by UmarJ** [comment]
Isn't it a problem that can occur anywhere two numbers are multiplied? I have no idea if it could even be fixed without modifying the multiplication function definition in CPython.

**Comment by zerothi** [comment]
Yes, but comparing two numbers with absolute difference does not require to multiply them, it is only required for some relative comparisons. However what does relative mean for infinite numbers in this case?

I would argue that the algorithm needs to be changed for elements containing inf. In this way the multiplication is circumvented for inf only.

**Comment by UmarJ** [comment]
My point is that there are many places in the code where two numbers are being multiplied. We can't really add that check for infinity everywhere. Besides, it's the underlying calculation which is incorrect, fixing which is better than circumventing the problem.

**Comment by zerothi** [comment]
See my comment about ieee-754 conformity in your linked bug report. Again, my point here is that one should be able to compare the corner cases strictly without resorting to multiplications since this is a comparison check, not a multiplication which seems to actually do the correct thing (except the real * complex values which one _could_ argue does the correct thing anyway ;) )

**Comment by UmarJ** [comment]
Is it really correct if it's mathematically incorrect however? I definitely do not get a `nan` when I solve it with a pen and paper or with sympy :)

**Comment by zerothi** [comment]
No, but IEEE 754 isn't pen and paper. It defines `0*inf=nan`.

**Comment by UmarJ** [comment]
That `0 * inf` calculation should not be happening in the first place is the entire point of the issue that I opened.

**Comment by sayantikabanik** [comment]
I gave a shot at this, checking whether the values are complex infinities using `cmath.isinf` (still looking at the big trail of failed tests). Though this might not be a full-fledged solution, as correctly pointed out adding this check everywhere isn't viable. 
<details>
<summary> trial </summary>
<img width="604" alt="Screenshot 2021-11-10 at 9 12 06 AM" src="https://user-images.githubusercontent.com/17350312/141045885-dca0dca9-fd1e-456c-bc1e-fe5e03efab42.png">
</details>

**Comment by andrewkwolek** [comment]
Checking for complex infinities could at least be a temporary patch for now
